### ISSUE TITLE
[SURE-4251] fleet.yaml vs GitRepo targeting rules

### ISSUE THREAD

**Issue by kkaempf** [rancher/fleet#1254]
While trying out different combinations of GitRepo targeting rules and fleet.yaml targetCustomizations, our team discovered some behavior that appears to suggest that the cluster selectors in both places must be linked a lot closer than the documentation would suggest.

-    Four downstream clusters: cluster1, cluster2, cluster3, and cluster4 (RKE deployed to AWS EC2 via the Rancher management cluster).
-    Four cluster groups: group1, ... group4 each with one corresponding cluster in them.
-    A fifth cluster group, "bens-clusters" with all four clusters in it.
-    The gitops repo itself has just a single fleet.yaml file in it. Containing only a release name and target customizations that specify a helm chart URL. Example attached.

Our expectation was that if we set the GitRepo to target the bens-clusters group, we would then be able to use groups 1-4 within the fleet.yaml to determine which version of a chart was deployed, or even choose not to deploy if one of the groups was not targeted.

What we've observed is that unless a group is included on the list of targets at the GitRepo level, it cannot be used in a targetCustomization rule within that repo. Fleet behaves as if that rule weren't there at all.

Specifically, we ran through the following combinations and noted the results.

1.
gitrepo target: bens-clusters
fleet targeting: bens-clusters
result: everybody gets same chart (expected)

2.
gitrepo target: bens-clusters
fleet targeting: group1, group2
result: no clusters get any deployments

3.
gitrepo target: bens-clusters
fleet targeting: group1, bens-clusters
result: everybody gets the same thing (even group1, their higher priority customization was ignored)

4.
gitrepo target: group1, bens-clusters
fleet targeting: group1, group2
result: only group1 gets anything

5.
gitrepo target: group1,2,3,4
fleet targeting: group1, group2
result: group1 and 2 each get the charts that match the customization rule, group 3 and 4 get nothing (also correct)

Scenario 1-3 show our desired use-case. A GitOps repo is configured to target a large number of downstream clusters, and the deployments can be managed entirely via updates to the fleet.yaml file(s) within the repo, without reconfiguring Rancher. Unfortunately in scenarios 2 and 3 we either don't get a deployment at all, or we release the wrong chart to group1.

Scenario 5 meets the same end-result as Scenario 2, in that we can deploy chartA to group1, chartB to group2, and nothing to groups 3 and 4 while still having the repo set up to target them. The downside of this approach is that any time we want to target a new grouping of downstream clusters, we must modify the GitRepo configuration to include that group in its target list, even if all the clusters within the group were already getting targeted.

So with all that said, my questions are.

1. Is this behavior intended, that we cannot reference groups in the fleet.yaml files if the repo is not explicitly configured to target that group?
2. If so, does this apply to other cluster selection mechanisms? e.g. if we want to customize based on cluster labels or individual clusters, must we include those as target rules in the repo configuration?
3. Could Fleet raise an error if it detects a targetCustomization that it can't apply due to the cluster group not being in the repo targets list?
4. Is there something we're doing wrong to run into this in the first place?
5. Any thoughts to nested cluster groups so that we could more easily manage large numbers of related clusters?

I recognize this is a rather complex question. Appreciate whatever information you can share.

**Comment by raulcabello** [comment]
I noticed some inconsistent behavior when there are more than one matches in both the `fleet.yaml` and `GitRepo`.  

Environment:
- Rancher 2.7.1
- Fleet v0.5.1
-  2 downstream clusers (down1 and down2)
-  3 `clustergroups`. One for each cluster and another that matches both.

When I create the following `GitRepo`:
```
kind: GitRepo
apiVersion: fleet.cattle.io/v1alpha1
metadata:
  # Any name can be used here
  name: nginx
  namespace: fleet-default
spec:
  repo: https://github.com/raulcabello/fleet-test
  branch: main
  paths:
  - nginx
  targets:
    - name: all
      clusterGroup: all
    - name: down1
      clusterGroup: down1
```
fleet.yaml
```
defaultNamespace: nginx
helm:
  releaseName: grafana
  chart: grafana
  repo: https://grafana.github.io/helm-charts
labels:
  key: valuetestchanged
targetCustomizations:
- name: down1
  clusterGroup: down1
  helm: 
    values:
      replicas: 3
- name: all
  clusterGroup: all
  helm: 
    values:
      replicas: 2
```
Sometimes I get the wrong replicas for `down1` and `down2` clusters, and if I recreate the `GitRepo` I get different replicas. It looks like a different `clusterGroup` is matched each time. 
The helm release values didn't match the `BundleDeployment` values in one occasion, which led to the following error in the `BundleDeployment`:
```
Modified(1) [Bundle nginx-nginx]; deployment.apps nginx/grafana modified {"spec":{"replicas":3}} 
```

**Comment by raulcabello** [comment]
## QA Template
 
### Problem
Target customizations are ignored if they are not defined in the `GitRepo`. For example:
- There is 2 `Clusters` and 3 `ClusterGroups`. There is one `ClusterGroup` that matches one of the clusters (one and two), and other that matches all.
- Create a `GitRepo` that targets `ClusterGroup` all
- Target customizations for `ClusterGroups` one and two don't work as they were not included in the `GitRepo`

 
### Solution
Check if there is any `TargetCustomization` that should be applied, even if it is not defined in the `GitRepo`. 
 
### Testing

### Engineering Testing
#### Manual Testing
The use cases defined in this issue should work. It can be checked with less than 4 clusters.
Steps:
- Create four downstream clusters: cluster1, cluster2, cluster3, and cluster4.
- Create four cluster groups: group1, ... group4 each with one corresponding cluster in them.
- Create a fifth cluster group, "all" with all four clusters in it.

Then verify that customizations targeting `ClusterGroups` that correspond to single clusters from the `targetCustomization` in the `fleet.yaml` are applied if we create a `GitRepo` that matches all targets. For this you can use any helm chart and modify the values in `targetCustomization` (or modify any other field instead of the values)

#### Automated Testing
[Integration tests](https://github.com/rancher/fleet/blob/master/integrationtests/controller/bundle/bundle_targets_test.go)

**Comment by markusewalker** [comment]
**QA TEST PLAN**

Scenarios
|Scenario|Test Case|
|---|---|
|1|As an admin user, provision four downstream node driver clusters, associate four cluster groups to each of them, a fifth one to all of the clusters and validate `targetCustomization` matches all targets|

**Comment by markusewalker** [comment]
Validated that this is addressed in `v2.7-head` https://github.com/rancher/rancher/commit/ed4b893709717f49c8c55cf9fe2399601007bcf8.

**ENVIRONMENT DETAILS**
- Rancher install: Helm
- Rancher version
     - Repro: `v2.7.1`
     - Validated: `v2.7-head` https://github.com/rancher/rancher/commit/ed4b893709717f49c8c55cf9fe2399601007bcf8 

**TEST RESULT**
|Scenario|Test Case|Result
|---|---|---|
|1|As an admin user, provision four downstream node driver clusters, associate four cluster groups to each of them, a fifth one to all of the clusters and validate `targetCustomization` matches all targets|:white_check_mark:|

**REPRO STEPS**

**Scenario 1**
1. As an admin user, provisioned four downstream K3S node driver clusters (mixture of v1.25.x+k3s1 and v1.24.x+k3s1).
2. Navigated to Continuous Delivery.
3. Created five cluster groups - 4 for each downstream node driver cluster, 1 that targets all of the downstream node driver clusters.
4. Created a Git Repo in the `fleet-default` namespace using the YAML defined in this comment.
5. Viewed the deployed bundle and validated that the `targetCustomization` and `targetRestrictions` are NOT correctly applied to the expected cluster groups.
     - The `targetRestrictions` are not properly being applied to the cluster groups at all in the `fleet.yaml` as mentioned in the issue.

**VALIDATION STEPS**

**Scenario 1**
1. As an admin user, provisioned four downstream K3S node driver clusters (mixture of v1.25.x+k3s1 and v1.24.x+k3s1).
2. Navigated to Continuous Delivery.
3. Created five cluster groups - 4 for each downstream node driver cluster, 1 that targets all of the downstream node driver clusters.
4. Created a Git Repo in the `fleet-default` namespace using the YAML defined in this comment.
5. Viewed the deployed bundle and validated that the `targetCustomization` and `targetRestrictions` is correctly applied to the expected cluster groups.
